### PR TITLE
feat(router): add an api to migrate the apple pay certificates from connector metadata to `connector_wallets_details` column in merchant connector account

### PR DESCRIPTION
**Pull request by ShankarSinghC**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Currently apple pay related details are being stored in the connector metadata, as in future we are going to add apple pay decrypted flow support for ios app we will be collecting the apple pay payment processing certificates as well. As this ppc and ppc key needs to be securely stored, we need to move the existing certificates to a new column which stores the encrypted certificates.

This pr adds a api end point (`/apple_pay_certificates_migration`) to migrate the apple pay details form the merchant connector account `metadata`  to the newly added column connector (`connector_wallets_details`). This api takes list of `merchant_ids` as input and then lists all the configured merchant connector accounts for that merchant id. After which it checks for the apple pay details in the metadata in the merchant connector account, if present it is encrypted by merchant's DEK (Data Encryption Key) and will be stored in the newly added column (`connector_wallets_details`) in the merchant connector account. Response contains the list of merchant_ids for migration succeeded (`migration_successful`) and failed (`migraiton_failed`).

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

-> Apple pay certificate migration api
```
curl --location 'http://localhost:8080/apple_pay_certificates_migration' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: test_admin' \
--data '{
    "merchant_ids": ["merchant_1713176513", "merchant_1714122879"]
}'
```
image
-> Check the db for connector_wallets_details of mca belonging to above merchants
image

-> Create MCA with apple pay `simplified` and manual `flow`
metadata for apple pay `simplified`
```
"apple_pay_combined": {
            "simplified": {
                "session_token_data": {
                    "initiative_context": "sdk-test-app.netlify.app",
                    "merchant_business_country": "US"
                },
                "payment_request_data": {
                    "label": "applepay",
                    "supported_networks": [
                        "visa",
                        "masterCard",
                        "amex",
                        "discover"
                    ],
                    "merchant_capabilities": [
                        "supports3DS"
                    ]
                }
            }
        }
```
metadata for apple pay `manual` flow
```
"apple_pay_combined": {
            "manual": {
                "session_token_data": {
                    "initiative": "web",
                    "certificate": "",
                    "display_name": "applepay",
                    "certificate_keys": "",
                    "merchant_business_coountry": "US"
                },
                "payment_request_data": {
                    "label": "applepay",
                    "supported_networks": [
                        "visa",
                        "masterCard",
                        "amex",
                        "discover"
                    ],
                    "merchant_capabilities": [
                        "supports3DS"
                    ]
                }
            }
        }
```
-> Confirm an apple pay payment with above flows
image
-> Create a apple pay payment with the below metadata
```
"apple_pay_combined": {
            "manual": {
                "session_token_data": {
                    "initiative": "web",
                    "certificate": "",
                    "display_name": "applepay",
                    "certificate_keys": "",
                    "payment_processing_details_at": "Hyperswitch",
                    "payment_processing_certificate": "",
                    "payment_processing_certificate_key": "",
                    "initiative_context": "sdk-test-app.netlify.app",
                    "merchant_identifier": "",
                    "merchant_business_coountry": "US"
                },
                "payment_request_data": {
                    "label": "applepay",
                    "supported_networks": [
                        "visa",
                        "masterCard",
                        "amex",
                        "discover"
                    ],
                    "merchant_capabilities": [
                        "supports3DS"
                    ]
                }
            }
        }
```
-> Confirm apple pay payment
image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payments/helpers.rs [hunk 1]
```diff
@@ -3931,6 +3955,38 @@ pub fn is_apple_pay_simplified_flow(
...
+pub async fn get_encrypted_apple_pay_connector_wallets_details(
+    key_store: &domain::MerchantKeyStore,
+    connector_metadata: &Option<masking::Secret<tera::Value>>,
+) -> RouterResult<Option<Encryptable<masking::Secret<serde_json::Value>>>> {
+    let apple_pay_metadata = get_applepay_metadata(connector_metadata.clone())
+        .map_err(|error| {
+            logger::info!(
+                "Apple pay metadata parsing failed in get_encrypted_apple_pay_connector_wallets_details {:?}",
+                error
+            )
```
Comment (SanchithHegde) [inline crates/router/src/core/payments/helpers.rs:3967]: Avoid info log lines for error messages.

File: crates/router/src/core/payments/helpers.rs [hunk 2]
```diff
@@ -3908,17 +3917,32 @@ pub fn validate_customer_access(
...
+            .map_err(|error| {
+                logger::info!(
+                "Apple pay metadata parsing failed for {:?} in is_apple_pay_simplified_flow {:?}",
                 connector_name,
                 error
             )
-        })
-        .ok();
+            })
+            .ok(),
```
Comment (SanchithHegde) [inline crates/router/src/core/payments/helpers.rs:3944]: Avoid info log lines for error messages.

File: crates/router/src/core/apple_pay_certificates_migration.rs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            .await;
+
+        match merchant_connector_accounts_update {
+            Ok(_) => {
+                migration_successful_merchant_ids.push(merchant_id.to_string());
+            }
+            Err(_) => {
+                migration_failed_merchant_ids.push(merchant_id.to_string());
+            }
+        };
```
Comment (SanchithHegde) [inline crates/router/src/core/apple_pay_certificates_migration.rs:98]: We can optionally add debug log lines in each case with suitable messages, as a sort of a progress indicator. Again, optional change.

**Review by SanchithHegde — APPROVED** [review]

**Review by ivor-juspay — APPROVED** [review]

**Review by jagan-jaya — APPROVED** [review]

**Review by Chethan-rao — APPROVED** [review]
@juspay/hyperswitch-routing related changes LGTM

**Pull request merged**